src/public_key/mod.rs

```diff
@@ -0,0 +1,673 @@
+{
+    type KeyPair = AbeKeyPair<T>;
+    type KeygenParam = (usize, String);
+
+    fn new() -> Self {
+        panic!("ABE scheme must be initialized with attributes.")
+    }
+
+    #[must_use]
+    fn new_attrs(attrs: &[u32]) -> Self {
+        Self {
+            scheme: T::default(),
+            attrs: Some(attrs.to_vec()),
+        }
+    }
+
+    fn description(&self) -> String {
+        format!("Abe {}", T::description())
+    }
+
+    fn generate_key_pair(&self, param: Self::KeygenParam) -> anyhow::Result<Self::KeyPair> {
+        // warning: in this case, the encryption key is the 'public key'
+        // and the decryption key is the 'private key'
+        let (master_priv_key, master_pub_key, _) = T::default().generate_master_key(param.0)?;
+        let user_decryption_key = T::default().key_generation(
+            &MonotoneSpanProgram::<i32>::parse(&param.1)?,
+            &master_priv_key,
+        )?;
+        Ok(Self::KeyPair {
+            public_key: AbePublicKey(master_pub_key),
+            private_key: user_decryption_key,
+        })
+    }
+
+    fn generate_random_bytes(&self, len: usize) -> Vec<u8> {
+        let rng = Mutex::new(CsRng::new());
+        let rng = &mut *rng.lock().expect("a mutex lock failed");
+        let mut bytes = vec![0_u8; len];
+        rng.fill_bytes(&mut bytes);
+        bytes
+    }
+
+    fn generate_symmetric_key<S: SymmetricCrypto>(
+        &self,
+        public_key: &<Self::KeyPair as KeyPair>::PublicKey,
+    ) -> anyhow::Result<(S::Key, Vec<u8>)> {
+        let random_plain = self.scheme.generate_random_plaintext()?;
+        let attrs = &self
+            .attrs
+            .clone()
+            .ok_or_else(|| anyhow::anyhow!("No attribute found for ABE scheme encryption"))?[..];
+        let ciphertext = self.scheme.encrypt(&random_plain, attrs, &public_key.0)?;
+        let hasher = Shake256::default();
+        let symkey = hasher
+            .chain(&random_plain.as_bytes()?)
+            .finalize_xof()
+            .read_boxed(S::Key::LENGTH)
+            .into_vec();
+        Ok((S::generate_key_from_rnd(&symkey)?, ciphertext.as_bytes()?))
+    }
+
+    fn decrypt_symmetric_key<S: SymmetricCrypto>(
+        &self,
+        private_key: &<Self::KeyPair as KeyPair>::PrivateKey,
+        data: &[u8],
+    ) -> anyhow::Result<S::Key> {
+        let decrypted = self.decrypt(private_key, data)?;
+        let hasher = Shake256::default();
+        let symkey = hasher
+            .chain(&decrypted)
+            .finalize_xof()
+            .read_boxed(S::Key::LENGTH)
+            .into_vec();
+        S::Key::parse(symkey)
+    }
+
+    fn encrypted_message_length(&self, _clear_text_message_length: usize) -> usize {
+        match self.attrs.clone() {
+            Some(attrs) => T::ciphertext_len(attrs.len()) as usize,
+            None => 0,
+        }
+    }
+
+    fn encrypt(&self, public_key: &AbePublicKey<T>, data: &[u8]) -> anyhow::Result<Vec<u8>> {
```

Comment:
Shouldn't be a real result instead of anyhow::Result ?

Reply:
Will be fixed later in https://github.com/Cosmian/abe_gpsw/issues/9